python_modules/dagster/dagster/_daemon/asset_daemon.py

```diff
@@ -241,50 +308,243 @@ def _run_iteration_impl(
+                            if sensor.get_current_instigator_state(
+                                all_auto_materialize_states.get(selector_id)
+                            ).is_running:
+                                sensors.append(sensor)
+
+            # TODO Adapt stored legacy cursor into per-evaluation-group cursors the first time AMP is run
+            # using evaluation groups
+        else:
+            sensors.append(
+                None  # Represents that there's a single set of ticks with no underlying sensor
+            )
+            all_auto_materialize_states = {}
+
+        for sensor in sensors:
+            selector_id = sensor.selector.get_id() if sensor else None
+
+            auto_materialize_state = all_auto_materialize_states.get(selector_id)
+
+            if sensor:
+                if not auto_materialize_state:
+                    assert sensor.default_status == DefaultSensorStatus.RUNNING
+                    auto_materialize_state = InstigatorState(
+                        sensor.get_external_origin(),
+                        InstigatorType.SENSOR,
+                        InstigatorStatus.AUTOMATICALLY_RUNNING,
+                        SensorInstigatorData(
+                            min_interval=sensor.min_interval_seconds,
+                            cursor=AssetDaemonCursor.empty().serialize(),
+                            last_sensor_start_timestamp=pendulum.now("UTC").timestamp(),
+                            sensor_type=SensorType.AUTOMATION_POLICY,
+                        ),
+                    )
+                    instance.add_instigator_state(auto_materialize_state)
+                elif is_under_min_interval(auto_materialize_state, sensor):
+                    continue
+            else:
+                if (
+                    self._last_pre_sensor_submit_time
+                    and now - self._last_pre_sensor_submit_time < self._pre_sensor_interval_seconds
+                ):
+                    # Wait until the pre-sensor interval has finished before running
+                    continue
+
+                self._last_pre_sensor_submit_time = now
+
+            yield from self._process_auto_materialize_tick_generator(
+                workspace_process_context,
+                sensor,
+                sensor_state_lock,
+                debug_crash_flags,
+            )
+
+    def _process_auto_materialize_tick(
+        self,
+        workspace_process_context: IWorkspaceProcessContext,
+        sensor: Optional[ExternalSensor],
+        sensor_state_lock: threading.Lock,
+        debug_crash_flags: SingleInstigatorDebugCrashFlags,
+    ):
+        return list(
+            self._process_auto_materialize_tick_generator(
+                workspace_process_context,
+                sensor,
+                sensor_state_lock,
+                debug_crash_flags,
+            )
+        )
+
+    def _process_auto_materialize_tick_generator(
```

Comment:
Are all the threading locks just in preparation when we submit this to a thread pool?  Everything in this PR is still single-threaded, yes?